Class/Practise 6.1/include/MessageDispatcher.hpp

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <functional>
+#include <string>
+#include <vector>
+#include <unordered_map>
+#include <mutex>
+#include <condition_variable>
+#include <queue>
+#include <variant>
+#include <thread>
+#include <print>
+
+// Define message types
+struct GreenMessage {
+    std::string text;
+    int counter;
+};
+
+struct BlueMessage {
+    double value1;
+    double value2;
+};
+
+struct OrangeMessage {
+    std::string text1;
+    std::string text2;
+    int number;
+    double value;
+};
+
+// Define the message variant type
+using Message = std::variant<GreenMessage, BlueMessage, OrangeMessage>;
+
+// Dispatcher class
+class MessageDispatcher {
+public:
+    // Singleton instance
+    static MessageDispatcher& getInstance();
+
+    // Publish methods
+    void publish(const GreenMessage& message);
+    void publish(const BlueMessage& message);
+    void publish(const OrangeMessage& message);
+
+    // Subscription methods
+    void subscribe(const std::function<void(const GreenMessage&)>& subscriber);
```

Comment:
Few notes here:
1. We can use using to make an alias for this signature like
using GreenMessageCb = std::function<void(const GreenMessage&)>;
2. It is better to have other name like subscription_cb or something like this cause the function itself is not a subscriber.
3. Minor: just an interesting topic to read. in case of std::function it is not always better to pass via const & - sometimes it is enough to pass by value. When we use lambda with such a call it will be converted to std::function so in such a case passing a value may allow us to avoid extra copy. But of course for small function like this it doesn't affect the program performance significantly.

Reply:
1. Created type aliases for the subscription callbacks using the using keyword for better readability and maintainability.
2. Changed the subscription methods to use the new type aliases. 
```cpp
void subscribe(GreenMessageCallback subscriber);
void subscribe(BlueMessageCallback subscriber);
void subscribe(OrangeMessageCallback subscriber);
```
3. The subscription callbacks are now passed by value instead of const & in MessageDispatcher.cpp